app.js

```diff
@@ -185,7 +186,48 @@ function MonitorApp ({
       });
     };
     
+    // Store in memcache db
+    // TODO: deprecate memcache db. just use mongo for everything.
     mjs.set(key, JSON.stringify(newRoutes), {}, handleErr);
+    
+    // Add new routes to mongo db log. Used for generating timeseries.
+    db.collection('routeLog').insertOne({
+      'timestamp': now,
+      // can omit timestamp from each route object since they're all the same
+      'routes': newRoutes.map((r) => _.omit(r, 'timestamp'))
+    });
+
+  }));
+
+  app.get('/api/v0/numNodesTimeseries', asyncMiddleware(async function(req, res) {
+    let numNodes = [];
+    let numGateways = [];
+    let timestamps = [];
+    
+    let now = new Date();
+    let yesterday = new Date(now - 1000 * 60 * 60 * 24);
+    let toDate = now;
+    let fromDate = yesterday;
+    if (req.query.from)
+      fromDate = new Date(req.query.from);
+    if (req.query.to)
+      toDate = new Date(req.query.to);
+
+    await db.collection('routeLog')
+      .find({
+        timestamp: {
+          '$lt': toDate,
+          '$gte': fromDate
+        }
+      })
+      .sort({ timestamp: 1 })
+      .forEach((routeLog) => {
+        timestamps.push(routeLog.timestamp);
+        numGateways.push(_.unique(routeLog.routes, (route) => route.gatewayIP).length);
```

Comment:
Ideally you'd ask Mongo to do this counting for you. That way you don't have to load the whole result set into memory here in node.js. When `routeLog.routes.length` is very big, the node.js process can crash if it runs out of memory.

You can push this to Mongo using the aggregation pipeline API: https://stackoverflow.com/questions/16368638/mongodb-distinct-aggregation

Reply:
Are you sure about this? I thought aggregation was about aggregating across documents (like in the example you linked).

I don't think we're ever expecting the number of routes to exceed node's available memory, so it's not a big deal, but I'd be curious to know what the mongo syntax for this would be.